### PR TITLE
feat(BrowserContext): introduce Browser Contexts.

### LINKED ISSUE #85: Support browser contexts to launch different sessions
Support browser contexts (Target.createBrowserContext) so to avoid launching multiple instances if one wants a pristine session. See proposal https://github.com/GoogleChrome/puppeteer/issues/66 and original discussion at https://github.com/cyrus-and/chrome-remote-interface/issues/118.

A viable user scenario might be testing several users logged in simultaneously into the service.
We might expose browser contexts as a string literal option to the browser.newPage:
```js
browser.newPage(); // creates a new page in a default browser context
browser.newPage({ context: 'default' }); // same as previous call
browser.newPage({ context: 'another-context' }); // creates a page in another browser context
```

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch introduces Browser Contexts and methods to manage them:
- `browser.createIncognitoBrowserContext()` - to create new incognito
  context
- `browser.browserContexts()` - to get all existing contexts
- `browserContext.dispose()` - to dispose incognito context.

Fixes #85.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -456,17 +465,41 @@ Emitted when a target is created, for example when a new page is opened by [`win
 
 Emitted when a target is destroyed, for example when a page is closed.
 
+#### browser.browserContexts()
+- returns: <[Array]<[BrowserContext]>>
+
+Returns all the browser contexts created in the browser. In a newly created browser, this will return
```
Comment (JoelEinbinder) [inline docs/api.md]: Returns an array of all open browser contexts. In a newly created browser, ...

File: docs/api.md [hunk 2]
```diff
@@ -456,17 +465,41 @@ Emitted when a target is created, for example when a new page is opened by [`win
...
+
 #### browser.close()
 - returns: <[Promise]>
 
 Closes Chromium and all of its pages (if any were opened). The [Browser] object itself is considered to be disposed and cannot be used anymore.
 
+#### browser.createIncognitoBrowserContext()
+- returns: <[Promise]<[BrowserContext]>>
+
+Creates a new incognito browser context. A typical use would be opening pages in a pristine
```
Comment (JoelEinbinder) [inline docs/api.md]: This won't share cookies/cache with other browser contexts.

File: docs/api.md [hunk 3]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
 
 You can find the `webSocketDebuggerUrl` from `http://${host}:${port}/json/version`. Learn more about the [devtools protocol](https://chromedevtools.github.io/devtools-protocol) and the [browser endpoint](https://chromedevtools.github.io/devtools-protocol/#how-do-i-access-the-browser-target).
 
+### class: BrowserContext
+
+BrowserContexts provide a way to operate multiple independent browser sessions. When browser is launched, it has
```
Comment (JoelEinbinder) [inline docs/api.md]: When *a* browser is launched...
The method ...

with the call `window.open`
or with a `window.open` call

the parent page's browser context

s/persist/write

File: docs/api.md [hunk 4]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+a single BrowserContext used by default. Method `browser.newPage()` creates a page in the default browser context.
+
+If a page opens another page, e.g. with the `window.open` call, the popup will belong to the parent page browser
+context.
+
+Puppeteer allows creation of "incognito" browser contexts with `browser.createIncognitoBrowserContext()` method.
+"Incognito" browser contexts don't persist any browsing data to disk.
+
+Typical usage of a context:
```
Comment (JoelEinbinder) [inline docs/api.md]: drop

File: docs/api.md [hunk 5]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+// ... do stuff with page ...
+await page.goto('https://example.com');
+// Dispose context once its no longer needed.
+await context.dispose();
+```
+
+#### browserContext.browser()
+- returns: <[Browser]>
+
+The browser the browser context belongs to.
```
Comment (JoelEinbinder) [inline docs/api.md]: The browser this browser context belongs to.

File: docs/api.md [hunk 6]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+
+#### browserContext.dispose()
```
Comment (JoelEinbinder) [inline docs/api.md]: .close?

File: docs/api.md [hunk 7]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+- returns: <[Promise]>
+
+Dispose the browser context. All the targets that belong to the browser context
+will be closed.
+
+> **NOTE** only non-incognito browser contexts can be disposed.
```
Comment (vsemozhetbyt) [inline docs/api.md]: This contradicts the example above and the note below.

File: docs/api.md [hunk 8]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+
+#### browserContext.isIncognito()
+- returns: <[boolean]>
+
+Returns whether BrowserContext is incognito. The only non-incognito context is a default browser context.
+
+> **NOTE** non-incognito browser contexts cannot be disposed.
```
Comment (JoelEinbinder) [inline docs/api.md]: the default browser context cannot be closed.

File: docs/api.md [hunk 9]
```diff
@@ -495,6 +531,61 @@ Browser websocket endpoint which can be used as an argument to
...
+
+#### browserContext.newPage()
+- returns: <[Promise]<[Page]>>
+
+Promise which resolves to a new [Page] object in this browser context.
```
Comment (JoelEinbinder) [inline docs/api.md]: Creates a new page in this browser context.

File: docs/api.md [hunk 10]
```diff
@@ -2607,6 +2698,12 @@ Contains the URL of the response.
 
 Get the browser the target belongs to.
 
+#### target.browserContext()
+
+- returns: <[BrowserContext]>
+
+Get the browser context the target belongs to.
```
Comment (JoelEinbinder) [inline docs/api.md]: The browser context this target belongs to.

File: docs/api.md (hunk 7 above)
Comment (JoelEinbinder) [reply docs/api.md]: Yes.

Comment (aslushnikov) [reply docs/api.md]: Yes.

File: docs/api.md (hunk 1 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 2 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 3 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 4 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 5 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 7 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 8 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 10 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 9 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

File: docs/api.md (hunk 6 above)
Comment (aslushnikov) [reply docs/api.md]: I like it. Done.

File: lib/Browser.js [hunk 11]
```diff
@@ -51,24 +59,53 @@ class Browser extends EventEmitter {
...
 
   /**
-   * @param {{targetInfo: !Puppeteer.TargetInfo}} event
+   * @param {!Protocol.Target.targetCreatedPayload} event
    */
   async _targetCreated(event) {
     const targetInfo = event.targetInfo;
-    const target = new Target(targetInfo, this, () => this._connection.createSession(targetInfo.targetId), this._ignoreHTTPSErrors, !this._appMode, this._screenshotTaskQueue);
+    const {browserContextId} = targetInfo;
+    const context = browserContextId && this._contexts.has(browserContextId) ? this._contexts.get(browserContextId) : this._defaultContext;
```
Comment (JoelEinbinder) [inline lib/Browser.js]: can you put paren around `browserContextId && this._contexts.has(browserContextId)`. I never remember order of operations ternary vs boolean operator.

**Comment by JoelEinbinder** [comment]
I think we should mirror the target events on the browser context. So I can await `targetcreated` on the browser context to find my popup without worrying about what other pages might be doing.

File: test/browsercontext.spec.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+const utils = require('./utils');
+
+module.exports.addTests = function({testRunner, expect, puppeteer}) {
+  const {describe, xdescribe, fdescribe} = testRunner;
+  const {it, fit, xit} = testRunner;
+  const {beforeAll, beforeEach, afterAll, afterEach} = testRunner;
+
+  describe('BrowserContext', function() {
+    it('should have default context', async function({browser, server}) {
```
Comment (JoelEinbinder) [inline test/browsercontext.spec.js:25]: Given the seperate implementations of browser contexts in headless and headful, id feel safer if we had a headful test here. Or maybe its time to run HEADLESS=false on one of the bots.

Comment (aslushnikov) [reply test/browsercontext.spec.js:25]: > Given the seperate implementations of browser contexts in headless and headful

This is not the first time we have separate implementations (e.g. `Target.createTarget`)

> Or maybe its time to run HEADLESS=false on one of the bots.

I'm all in for this. IIRC there's some work to be done upstream to make it possible. Filed #2524.

**Comment by aslushnikov** [comment]
> I think we should mirror the target events on the browser context. So I can await targetcreated on the browser context to find my popup without worrying about what other pages might be doing.

Valid usecase. Let me add these.

File: lib/Browser.js (hunk 11 above)
Comment (aslushnikov) [reply lib/Browser.js]: Done. 

> I never remember order of operations ternary vs boolean operator

Interestingly I had the same issue, and eventually found that dealing with these kind of expressions in codebases help me memorize the rules. For example, there's a common pattern to handle nullable collections:

```js
const result = arr && arr.includes(value) ? 'yeah!' : 'nope.';
```

**Review by JoelEinbinder — APPROVED** [review]

**Comment by zhaopengme** [comment]
good news!

**Pull request merged**